src/python/tests/gff3/test_restructure.py

```diff
@@ -0,0 +1,314 @@
+    [
+        param(["mRNA"], ["mRNA"], does_not_raise(), id="mRNA only, skip"),
+        param(["CDS"], ["CDS"], does_not_raise(), id="1 CDS, skip"),
+        param(["CDS", "exon"], ["CDS", "exon"], does_not_raise(), id="CDS + exon, skip"),
+        param(["exon"], [{"mRNA": ["exon"]}], does_not_raise(), id="1 exon moved"),
+        param(["exon", "exon"], [{"mRNA": ["exon", "exon"]}], does_not_raise(), id="2 exons moved"),
+    ],
+)
+def test_move_only_exons_to_new_mrna(
+    children: List[str],
+    expected_children: Dict[str, int],
+    expectation: ContextManager,
+):
+    """Test the creation of a new mRNA for exons under a gene."""
+    gen = FeatGenerator()
+    gene = gen.make("gene", 1)[0]
+    gene.sub_features += gen.make_structure(children)
+    with expectation:
+        move_only_exons_to_new_mrna(gene)
+        assert gen.get_sub_structure(gene) == {"gene": expected_children}
+
+
+@pytest.mark.parametrize(
+    "children, diff_exon, expected_children, expectation",
+    [
+        param(["mRNA"], False, ["mRNA"], does_not_raise(), id="mRNA only, skip"),
+        param(["mRNA", "mRNA"], False, ["mRNA", "mRNA"], does_not_raise(), id="2 mRNA only, skip"),
+        param(["CDS"], False, ["CDS"], does_not_raise(), id="One CDS, skip"),
+        param(["CDS", "exon"], False, ["CDS", "exon"], does_not_raise(), id="CDS+exon, skip"),
+        param(["mRNA", "CDS"], False, [{"mRNA": ["exon", "CDS"]}], does_not_raise(), id="1 mRNA + 1 CDS"),
+        param(
+            ["mRNA", "CDS", "extra"],
+            False,
+            ["extra", {"mRNA": ["exon", "CDS"]}],
+            does_not_raise(),
+            id="1 mRNA + 1 CDS + extra",
+        ),
+        param(
+            ["mRNA", "CDS", "CDS"],
+            False,
+            [{"mRNA": ["exon", "exon", "CDS", "CDS"]}],
+            does_not_raise(),
+            id="1 mRNA + 2 CDS",
+        ),
+        param(["mRNA", "mRNA", "CDS"], False, [], raises(GFFParserError), id="2 mRNA only + CDS, fail"),
+        param(
+            ["mRNA", "mRNA", "CDS", "exon"],
+            False,
+            [],
+            raises(GFFParserError),
+            id="2 mRNA only + CDS + exon, fail",
+        ),
+        param([{"mRNA": ["CDS"]}, "CDS"], False, [], raises(GFFParserError), id="1 mRNA/CDS + 1 CDS, fail"),
+        param(
+            [{"mRNA": ["exon"]}, "CDS"],
+            False,
+            [{"mRNA": ["exon", "CDS"]}],
+            does_not_raise(),
+            id="1 mRNA/exon + 1 CDS same",
+        ),
+        param([{"mRNA": ["exon"]}, "CDS"], True, [], raises(GFFParserError), id="1 mRNA/exon + 1 CDS diff"),
+        param(
+            [{"mRNA": ["extra"]}, "CDS"],
+            False,
+            [{"mRNA": ["extra", "exon", "CDS"]}],
+            does_not_raise(),
+            id="1 mRNA/extra + 1 CDS",
+        ),
+        param(
+            [{"mRNA": ["exon", "exon"]}, "CDS"],
+            True,
+            [],
+            raises(GFFParserError),
```

Comment:
In this case you may want to add the `match=...` argument to `raises()` [[link](https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions)] to make sure the correct `GFFParseError` is raised, as there are 2 conditions for which this exception is raised in the function.

Reply:
Hm that's a bit more work, but I guess that'd make sure we raise the exception where we expect it to...